synapse/storage/databases/main/events_worker.py

```diff
@@ -729,12 +731,16 @@ async def get_missing_events_from_db() -> Dict[str, EventCacheEntry]:
 
         return event_entry_map
 
-    def _invalidate_get_event_cache(self, event_id: str) -> None:
-        self._get_event_cache.invalidate((event_id,))
+    async def _invalidate_get_event_cache(self, event_id: str) -> None:
+        await self._get_event_cache.invalidate((event_id,))
+        self._invalidate_local_get_event_cache(event_id)
+
+    def _invalidate_local_get_event_cache(self, event_id: str) -> None:
+        self._get_event_cache.lru_cache.invalidate((event_id,))
```

Comment:
This feels like it's poking into the implementation of `AsyncLruCache`. It might be better to give `AsyncLruCache` an `invalidate_local_only` method or something.

Reply:
Added in https://github.com/matrix-org/synapse/pull/13242/commits/037d84f68418504a147b0114f9a905f9e2c369a1